File: frontend/src/types/domain.ts
export type AlertStatus = "new" | "in_progress" | "resolved";

export type Evidence = {
  feature: string;
  value: string;
  importance: number;
};

export type NormalizedEvent = {
  eventID: string;
  eventName: string;
  eventTime: string;
  userIdentity: Record<string, any>;
  principal: string;
  sourceIPAddress: string;
  userAgent: string;
  awsRegion: string;
  eventSource: string;
  eventStatus: "Success" | "Error";
  risk_score: number;
  risk_level: string;
  requestParameters?: Record<string, any> | null;
  responseElements?: any;
  anomaly_reason?: string | null;
  raw?: Record<string, any>;
};

export type Alert = {
  alert_id: string;
  user_id: string;
  detection_method: "rule" | "ai";
  rule_id?: string;
  rule_name?: string;
  risk_score: number;
  status: AlertStatus;
  detected_at: string;
  summary: string;
  related_events: string[];
  evidence: Evidence[];
  suppressed?: boolean;
};

export type DashboardSummary = {
  risk_score: { current: number; previous: number; trend: string };
  risk_trend: Array<{ time: string; risk: number; events: number }>;
  event_histogram: Array<{ time: string; count: number }>;
  top_anomalies: Alert[];
  signal_distribution: Array<{ service: string; value: number }>;
  totals: { events: number; alerts: number; principals: number; source_ips: number };
};

export type TimelineEvent = {
  id: string;
  time: string;
  timestamp: string;
  event_name: string;
  status: string;
  risk_score: number;
  is_anomaly: boolean;
  stage: string;
  anomaly_reason?: string | null;
};

export type AttackGraph = {
  alert_id: string;
  nodes: Array<{ id: string; type: string; label: string; risk_score: number; arn?: string | null }>;
  edges: Array<{ source: string; target: string; action: string; timestamp: string; risk: string; event_id: string }>;
  kill_chain: Array<{ stage: string; events: string[]; timestamp: string; description: string; risk: string }>;
};

File: backend/app/schemas/domain.py
from typing import Any, Literal
from pydantic import BaseModel, Field


class Evidence(BaseModel):
    feature: str
    value: str
    importance: float = Field(ge=0, le=1)


class NormalizedEvent(BaseModel):
    eventID: str
    eventName: str
    eventTime: str
    userIdentity: dict[str, Any] = Field(default_factory=dict)
    principal: str = "unknown"
    sourceIPAddress: str = "unknown"
    userAgent: str = "unknown"
    awsRegion: str = "unknown"
    eventSource: str = "unknown"
    eventStatus: Literal["Success", "Error"] = "Success"
    risk_score: int = Field(ge=0, le=100)
    risk_level: Literal["low", "medium", "high", "critical"]
    requestParameters: dict[str, Any] | None = None
    responseElements: Any = None
    anomaly_reason: str | None = None
    raw: dict[str, Any] = Field(default_factory=dict)


class Alert(BaseModel):
    alert_id: str
    user_id: str
    detection_method: Literal["rule", "ai"] = "rule"
    rule_id: str | None = None
    rule_name: str | None = None
    risk_score: int = Field(ge=0, le=100)
    status: Literal["new", "in_progress", "resolved"] = "new"
    detected_at: str
    summary: str
    related_events: list[str] = Field(default_factory=list)
    evidence: list[Evidence] = Field(default_factory=list)
    suppressed: bool = False


class AlertStatusUpdate(BaseModel):
    status: Literal["new", "in_progress", "resolved"]


class TimelineEvent(BaseModel):
    id: str
    time: str
    timestamp: str
    event_name: str
    status: str
    risk_score: int
    is_anomaly: bool
    stage: str
    anomaly_reason: str | None = None


class GraphNode(BaseModel):
    id: str
    type: str
    label: str
    risk_score: int
    arn: str | None = None


class GraphEdge(BaseModel):
    source: str
    target: str
    action: str
    timestamp: str
    risk: str
    event_id: str


class KillChainStep(BaseModel):
    stage: str
    events: list[str]
    timestamp: str
    description: str
    risk: str
